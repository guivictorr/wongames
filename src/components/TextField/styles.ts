import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from '.'

type InputWrapperProps = Pick<TextFieldProps, 'iconPosition'>

export const InputWrapper = styled.div<InputWrapperProps>`
  ${({ theme, iconPosition = 'left' }) => css`
    display: flex;
    flex-direction: ${iconPosition === 'right' && 'row-reverse'};
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`

export const Icon = styled.div`
  ${({ theme }) => css`
    display: flex;
    width: 2.2rem;
    color: ${theme.colors.gray};
    & > svg {
      width: 100%;
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};
      &::placeholder {
        color: currentColor;
      }
    }
  `
}

export const Wrapper = styled.div<Pick<TextFieldProps, 'disabled'>>`
  ${({ theme, disabled }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`
