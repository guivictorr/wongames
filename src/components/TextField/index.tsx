import { ChangeEvent, InputHTMLAttributes, useState } from 'react'
import * as S from './styles'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  icon?: JSX.Element
  initialValue?: string
  iconPosition?: 'left' | 'right'
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  initialValue,
  label = '',
  labelFor = '',
  onInput,
  icon,
  iconPosition,
  disabled,
  ...rest
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper disabled={disabled}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper iconPosition={iconPosition}>
        {!!icon && <S.Icon>{icon}</S.Icon>}
        <S.Input
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          {...rest}
        />
      </S.InputWrapper>
    </S.Wrapper>
  )
}

export default TextField
