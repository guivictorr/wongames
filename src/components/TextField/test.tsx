import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'

describe('<TextField />', () => {
  it('should render with label', () => {
    renderWithTheme(<TextField label="Label" labelFor="field" id="field" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('should render without label', () => {
    renderWithTheme(<TextField />)
    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="hey you" />)
    expect(screen.getByPlaceholderText(/hey you/i)).toBeInTheDocument()
  })

  it('should change the value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="Label"
        labelFor="textfield"
        id="textfield"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This text should be written'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })

    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('should be accessible by tab', () => {
    renderWithTheme(
      <TextField labelFor="textfield" label="Textfield" id="textfield" />
    )
    expect(document.body).toHaveFocus()
    userEvent.tab()
    expect(screen.getByRole('textbox')).toHaveFocus()
  })
})
