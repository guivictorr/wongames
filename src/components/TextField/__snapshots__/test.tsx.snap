// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render error message and change style 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c3:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c4 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c1 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c0 .c2 {
  border-color: #FF6347;
}

.c5 {
  color: #FF6347;
}

<div
  class="c0"
  disabled=""
>
  <label
    class="c1"
    for="textfield"
  >
    Textfield
  </label>
  <div
    class="c2 c3"
  >
    <input
      class="c4"
      disabled=""
      id="textfield"
      type="text"
      value=""
    />
  </div>
  <p
    class="c5"
  >
    Error message
  </p>
</div>
`
